### PR TITLE
fix: use existing rather than deleted models

### LINKED ISSUE #9553: StreamARN to resource is not cleaned up after type was removed from schema
### Before opening, please confirm:

- [X] I have installed the latest version of the Amplify CLI (see above), and confirmed that the issue still persists.
- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-cli/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### How did you install the Amplify CLI?

npm

### If applicable, what version of Node.js are you using?

12

### Amplify CLI Version

7.6.5

### What operating system are you using?

MacOS

### Did you make any manual changes to the cloud resources managed by Amplify? Please describe the changes made.

No

### Amplify Categories

function, api

### Amplify Commands

push

### Describe the bug

After removing a `@model` type from the schema that was referenced by a `function` resource `StreamARN` will be kept around and push is failing, dependency is not cleaned up.

export/getatt references are not removed from Lambda function and the tableA delete will go in "Update_rollback_complete"

### Expected behavior

During GQL compile a warning is printed about the reference and properly remove related CRUD resources dependencies and push should succeed

### Reproduction steps

1. Create Amplify App 
2. Add  GraphQL API with two models(tableA , tableB)
3 Add Lambda function with trigger template and add the trigger to one of the model(tableA) of the API
4 remove the tableA from the schema.graphql file

### GraphQL schema(s)

<details>

```graphql
# Put schemas below this line


```

</details>

### Log output

<details>

```
# Put your logs below this line


```

</details>

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by marcvberg**
#### Description of changes
When looking for missing function dependencies on API, a model diff against current backend doesn't work if the model was both added, function added with dependency, and model subsequently removed prior to push. This addresses that by looking to see if functions depend on models not in the backend, rather than diffing against cloud backend

#### Issue #, if available
#9553 

#### Description of how you validated changes
Running unit tests, making an application and testing using the original reproduction steps

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-category-function/src/provider-utils/awscloudformation/utils/getDependentFunction.ts [hunk 1]
```diff
@@ -1,9 +1,12 @@
...
 import * as path from 'path';
-import { loadFunctionParameters } from './loadFunctionParameters';
+import { loadFunctionParameters, loadFunctionStackAsJSON } from './loadFunctionParameters';
 import { ServiceName } from './constants';
 import { categoryName } from '../../../constants';
 
+/**
+ * @deprecated in favor of lambdasWithMissingApiDependency, using a list of existing models rather than deleted ones
+ */
 export async function lambdasWithApiDependency(
```
Comment (jhockett) [inline packages/amplify-category-function/src/provider-utils/awscloudformation/utils/getDependentFunction.ts:10]: Can this be removed now?

File: packages/amplify-category-function/src/provider-utils/awscloudformation/utils/getDependentFunction.ts [hunk 2]
```diff
@@ -38,3 +41,53 @@ export async function lambdasWithApiDependency(
...
+        }
+      }
+    }
+    else {
+      // In the case that the function-parameters.json does not have info on dependencies, we check the CFN stack.
+      // If the dependency is only a dynamo stream trigger, we don't have info in function-parameters.json to work with
+      const lambdaStackJson = loadFunctionStackAsJSON(resourceDirPath, lambda.resourceName);
+      if(lambdaStackJson?.Resources) {
+        const triggerPolicyTables = Object.keys(lambdaStackJson.Resources).filter(key => key.startsWith("LambdaTriggerPolicy"));
+        for(let triggerPolicy of triggerPolicyTables) {
```
Comment (jhockett) [inline packages/amplify-category-function/src/provider-utils/awscloudformation/utils/getDependentFunction.ts:83]: nit: triggerPolicy can be const
```suggestion
        for(const triggerPolicy of triggerPolicyTables) {
```

File: packages/amplify-category-function/src/provider-utils/awscloudformation/utils/updateDependentFunctionCfn.ts [hunk 3]
```diff
@@ -105,4 +110,40 @@ export function addAppSyncInvokeMethodPermission(
...
+  const functionCfn = JSONUtilities.readJson<$TSAny>(cfnFilePath);
+  let tableDeletions: string[] = [];
+  if(functionCfn?.Resources) {
+    const triggerPolicyTables = Object.keys(functionCfn.Resources).filter(key => key.startsWith('LambdaTriggerPolicy'));
+    for(let triggerPolicy of triggerPolicyTables) {
+      const tableName = triggerPolicy.match(/(?<=LambdaTriggerPolicy)(.*)/g)?.[0];
+      if(!sanitizedExistingModels.includes(tableName)) {
+        const triggerPolicyStatement: $TSAny = functionCfn?.Resources?.[triggerPolicy]?.Properties?.PolicyDocument?.
+          Statement;
+        for(let statement of triggerPolicyStatement ?? []) {
```
Comment (jhockett) [inline packages/amplify-category-function/src/provider-utils/awscloudformation/utils/updateDependentFunctionCfn.ts:133]:
```suggestion
        for(const statement of triggerPolicyStatement ?? []) {
```

**Review by jhockett — APPROVED** [review]
LGTM

File: packages/amplify-category-function/src/provider-utils/awscloudformation/utils/getDependentFunction.ts (hunk 1 above)
Comment (marcvberg) [reply packages/amplify-category-function/src/provider-utils/awscloudformation/utils/getDependentFunction.ts:10]: Yes, deleting this just forgot to delete it after removing references

File: packages/amplify-category-function/src/provider-utils/awscloudformation/utils/getDependentFunction.ts [hunk 4]
```diff
@@ -29,12 +31,26 @@ export async function lambdasWithApiDependency(
 
     if (typeof selectedCategories === 'object' && selectedCategories !== null) {
       for (const selectedResources of Object.values(selectedCategories)) {
-        deletedModelFound = Object.keys(selectedResources).some(r => modelsDeleted.includes(r));
+        deletedModelFound = Object.keys(selectedResources).some(r => !(existingModels.includes(r)));
```
Comment (cjihrig) [inline packages/amplify-category-function/src/provider-utils/awscloudformation/utils/getDependentFunction.ts:34]: Are the extra parentheses needed here?

File: packages/amplify-category-function/src/provider-utils/awscloudformation/utils/getDependentFunction.ts [hunk 5]
```diff
@@ -29,12 +31,26 @@ export async function lambdasWithApiDependency(
...
         if (deletedModelFound) {
           dependentFunctions.push(lambda);
         }
       }
     }
+    else {
+      // In the case that the function-parameters.json does not have info on dependencies, we check the CFN stack.
+      // If the dependency is only a dynamo stream trigger, we don't have info in function-parameters.json to work with
+      const lambdaStackJson = loadFunctionStackAsJSON(resourceDirPath, lambda.resourceName);
+      if(lambdaStackJson?.Resources) {
```
Comment (cjihrig) [inline packages/amplify-category-function/src/provider-utils/awscloudformation/utils/getDependentFunction.ts:44]: It doesn't look like the linter was run on these changes.

File: packages/amplify-category-function/src/provider-utils/awscloudformation/utils/getDependentFunction.ts [hunk 6]
```diff
@@ -29,12 +31,26 @@ export async function lambdasWithApiDependency(
...
+        const triggerPolicyTables = Object.keys(lambdaStackJson.Resources).filter(key => key.startsWith("LambdaTriggerPolicy"));
+        for(const triggerPolicy of triggerPolicyTables) {
+          const tableName = triggerPolicy.match(/(?<=LambdaTriggerPolicy)(.*)/g)?.[0];
+          if(!existingModels.includes(tableName)) {
```
Comment (cjihrig) [inline packages/amplify-category-function/src/provider-utils/awscloudformation/utils/getDependentFunction.ts:48]:
```suggestion
          if (tableName && !existingModels.includes(tableName)) {
```

File: packages/amplify-category-function/src/provider-utils/awscloudformation/utils/updateDependentFunctionCfn.ts [hunk 7]
```diff
@@ -105,4 +110,40 @@ export function addAppSyncInvokeMethodPermission(
...
+    return value.slice(0, value.indexOf(':'));
+  });
+  const cfnFilePath = path.join(resourceDirPath, `${resourceName}-cloudformation-template.json`);
+  const functionCfn = JSONUtilities.readJson<$TSAny>(cfnFilePath);
+  let tableDeletions: string[] = [];
+  if(functionCfn?.Resources) {
+    const triggerPolicyTables = Object.keys(functionCfn.Resources).filter(key => key.startsWith('LambdaTriggerPolicy'));
+    for(let triggerPolicy of triggerPolicyTables) {
+      const tableName = triggerPolicy.match(/(?<=LambdaTriggerPolicy)(.*)/g)?.[0];
+      if(!sanitizedExistingModels.includes(tableName)) {
```
Comment (cjihrig) [inline packages/amplify-category-function/src/provider-utils/awscloudformation/utils/updateDependentFunctionCfn.ts:130]:
```suggestion
      if (tableName && !sanitizedExistingModels.includes(tableName)) {
```

**Review by cjihrig — COMMENTED** [review]
Left some small comments. Mostly LGTM. Please run the linter against these changes.

File: packages/amplify-category-function/src/provider-utils/awscloudformation/utils/getDependentFunction.ts (hunk 5 above)
Comment (marcvberg) [reply packages/amplify-category-function/src/provider-utils/awscloudformation/utils/getDependentFunction.ts:44]: The linter isn't giving me any concerns with these lines, is there something specific that looks stylistically wrong?

Comment (jhockett) [reply packages/amplify-category-function/src/provider-utils/awscloudformation/utils/getDependentFunction.ts:44]: Prettier should add a space in the if statement: `if (...)`

Comment (cjihrig) [reply packages/amplify-category-function/src/provider-utils/awscloudformation/utils/getDependentFunction.ts:44]: Yea, what @jhockett said. I'm not sure what's going on with your local setup, but it keeps letting these style issues creep up in PRs.

**Review by cjihrig — APPROVED** [review]

**Pull request merged**